### ISSUE TITLE
Running out of memory during findChromPeaks

### ISSUE THREAD

**Issue by karlfraser** [sneumann/xcms#256]
Hi,

I'm trying to run a large-ish experiment using the new xcms3, but have hit some memory issues.
I have successfully processed this data already using an older version of xcms, so am a little surprised that this failed.  I am wanting to compare the results of the older version of xcms to xcms3.

Simply, 920 samples run as 8 batches on an Exactive at 25000 resolution and an 18 min chromatogram, converted using ProteoWizard to mzML format.  Data has been structured in Blanks (~40 files), QC (~90 files), and Samples (~790 files) and from that I have created using readMSData a OnDiskMSnExp object of 4705478 elements that is 1006Mb in size (took almost an hour to create).
The next step of findChromPeaks ran for approximately an hour and then gave a memory error.  code and error below:

# generate XCMSnExp object 
> raw_data <- readMSData(mzMLfiles, msLevel. = 1, mode = "onDisk")
> # ==HPLC  Orbitrap 
> cwp <- CentWaveParam(ppm = 15, mzdiff = -0.001, peakwidth =c (10,65), prefilter = c(3,3000), snthresh = 10, integrate = 1, noise = 1000)
> foundPeaks <- findChromPeaks(raw_data, param = cwp)
**Error: 'bplapply' receive data failed:
  memory exhausted (limit reached?)**
> 

The PC is a Z440 workstation with 64Gb of RAM running win7, so I'm surprised I ran out of memory.
Packages used to get things going at the start:
#=== load required R packages ================================
library(xcms)
library(CAMERA)
library(snow)
library(multtest)
library(Biobase)
library(purrr)
#library(data.table)
library(RColorBrewer)
library(BiocParallel)
register(SnowParam(16))

Any help would be very much appreciated.  Are there better ways of using the memory?  I assumed using the 'mode = onDisk' command that it is using the disk, not loading all files into memory, but I'm not sure that is happening.

Best Regards
Karl Fraser

PS: system info is:

print(sessionInfo())
R version 3.4.2 (2017-09-28)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=English_New Zealand.1252 LC_CTYPE=English_New Zealand.1252 LC_MONETARY=English_New Zealand.1252
[4] LC_NUMERIC=C LC_TIME=English_New Zealand.1252

attached base packages:
[1] parallel stats graphics grDevices utils datasets methods base

other attached packages:
[1] multtest_2.34.0 snow_0.4-2 CAMERA_1.34.0 xcms_3.0.0 MSnbase_2.4.0 ProtGenerics_1.10.0
[7] mzR_2.12.0 Rcpp_0.12.13 BiocParallel_1.12.0 Biobase_2.38.0 BiocGenerics_0.24.0

loaded via a namespace (and not attached):
[1] splines_3.4.2 lattice_0.20-35 colorspace_1.3-2 htmltools_0.3.6 stats4_3.4.2
[6] base64enc_0.1-3 vsn_3.46.0 XML_3.98-1.9 survival_2.41-3 RBGL_1.54.0
[11] rlang_0.1.2 foreign_0.8-69 affy_1.56.0 RColorBrewer_1.1-2 affyio_1.48.0
[16] foreach_1.4.3 plyr_1.8.4 stringr_1.2.0 mzID_1.16.0 zlibbioc_1.24.0
[21] munsell_0.4.3 pcaMethods_1.70.0 gtable_0.2.0 htmlwidgets_0.9 codetools_0.2-15
[26] knitr_1.17 latticeExtra_0.6-28 IRanges_2.12.0 doParallel_1.0.11 BiocInstaller_1.28.0
[31] MassSpecWavelet_1.44.0 preprocessCore_1.40.0 htmlTable_1.9 acepack_1.4.1 backports_1.1.1
[36] checkmate_1.8.5 scales_0.5.0 limma_3.34.0 S4Vectors_0.16.0 Hmisc_4.0-3
[41] graph_1.56.0 gridExtra_2.3 RANN_2.5.1 impute_1.52.0 ggplot2_2.2.1
[46] digest_0.6.12 stringi_1.1.5 grid_3.4.2 tools_3.4.2 magrittr_1.5
[51] lazyeval_0.2.1 tibble_1.3.4 Formula_1.2-2 cluster_2.0.6 pkgconfig_2.0.1
[56] MASS_7.3-47 Matrix_1.2-11 data.table_1.10.4-3 iterators_1.0.8 MALDIquant_1.16.4
[61] rpart_4.1-11 igraph_1.1.2 nnet_7.3-12 compiler_3.4.2

**Comment by jotsetung** [comment]
Hi Karl,
sorry to hear that you ran out of memory. In principle you should be able to run this without any problems. The `onDisk` mode is definitely using less memory, since all of the data is read *on-the-fly* from the original mzML files. So, in your case, the `findChromPeaks` call will read the full data of 16 files in parallel into memory to process them. I don't know how big the data of one of your files is, but you'll need 16 times that memory.

Were you able to process 16 files in parallel with the old `xcms`?

For our large experiments I tend to reduce the number of files being processed in parallel. Reason is that while you might have 16 CPUs, you still have a single I/O line/hard disk. So I/O will be a bottleneck for the parallel processing. I'd reduce the number of parallel processes and retry.

Another way to reduce memory demand is if your mzML files contain more data than you need. For our LCMS setup we know that after a certain time there is no interesting signal anymore coming through the LC. Our mzML files contain the full rt range, but we subset the object/data prior to peak detection.
Plotting a BPC on a representative file of yours might help:

```r
chr <- chromatogram(filterFile(raw_data, file = 1), aggregationFun = "max")
plot(chr)
``` 

Say you see a signal up to 400 seconds and nothing interesting anymore after that you could do:
```r
raw_data <- filterRt(raw_data, rt = c(0, 400))
```

**Comment by karlfraser** [comment]
Thanks very much for the advice.  Should have probably mentioned earlier but the file size of my 920 mzML files is between 65-80Mb each.

I have spent the last couple of days trying two different settings, sadly with no luck.
After rebooting the PC, the first attempt was to reduce the SnowParam to 4 (instead of 16). However, this again failed, with the memory slowly filling to 64Gb and then eventually failed leaving the following error:
foundPeaks <- findChromPeaks(raw_data, param = cwp)
Error: 'bplapply' receive data failed:
  cannot allocate vector of size 5 Kb

So, I rebooted and this time excluded the Snow librabry and Snow Param settings completely, so it would run on a single core.  Again the memory slowly filled and eventually failed leaving the same erro:
foundPeaks <- findChromPeaks(raw_data, param = cwp)
Error: 'bplapply' receive data failed:
  cannot allocate vector of size 6 Kb

I know my script works for small sample sets (we have got this working on 40 odd samples several times) and I have tested it on 115 samples successfully as well (and each of those mzML files were ~120Mb).  So I'm starting to think how can I get this to work now, and your suggestion of reducing file size is logical.  I have already cut off the front and rear of the chromatograms using MSConvert, but I'm wondering now about other settings I can use for MSConvert to reduce file size. I know if we convert as mzXML 32 bit instead of mzXL 64 bit (and I always use zlib compression), the files are roughly 1/2 the size, so even though I have accurate mass Exactive data this might be a sensible option (we used to use 32 bit before and got sensible results).

Do you have any other ideas about what could be going on with memory management and this step failing, or any other suggestions about how we might produce smaller files with MSConvert?

I potentially have larger experiments to process than this one...

Best Regards
Karl Fraser

**Comment by stanstrup** [comment]
How far does it get? As soon as it has loaded the raw data for all the parallel threads or only after processing many files? It is unusual to run out of memory at the peak picking step to me. I would keep an eye on the process monitor to understand how much each thread is using.

If you process a few files do you get a sane number of features? If the answer to the last is no then I would check if your noise and prefilter settings could reasonably be more conservative.

Probably @jotsetung knows better but I would imagine that reducing the bits in the raw files won't make much of a difference memory-wise since they will be read into the same format in R anyway.

**Comment by jotsetung** [comment]
Thanks for the feedback (also to @stanstrup for chiming in). Reducing the file size of the mzML files does not help. Whatever precision you have in the file, it will nevertheless get read as a `numeric` (i.e. double) in `R`.

From what I understand you see the memory slowly filling up. This could be due:
+ the `matrix` with the identified chromatographic peaks increases with each file.
+ the allocated memory is not completely freed/returned.

There is not much that can be done for the first point (unless you play around with the settings as suggested by @stanstrup). The second one would be more problematic. This would mean that memory is leaked. Every now and then I hear complains about memory problems on Windows, so it *could* be that there is something going on in Windows. Do you have any chance to try running the code on a linux or mac?

**Comment by jotsetung** [comment]
Hi Karl,

I made a small change that could eventually help to reduce memory demand a bit. Could you please try it?

To install:
```r
library(devtools)
install_github("sneumann/xcms", ref = "force_gc")
```

Let me know if this installation procedure does not work for you - I will then try to compile the package for Windows.

**Comment by karlfraser** [comment]
Hi,

That has fixed it!  Great stuff.  Thank you very much. Let me walk you through what I did, in case it may mean something to you for the future.
So, I went back to basics first, and did 115 files.  I noticed that after the Msn object had been made, when I started the findChromPeaks step the memory would continue to increase to a point, then dip and increase again to a point and then reamain at that level until the process had finished.  When i was doing 920 files, it would never get to this dip point.
![image](https://user-images.githubusercontent.com/34637659/36878568-72ee1f8a-1e24-11e8-9964-ed17243285e4.png)
It worked fine with 115 files (and snow at 16 cause I was in a hurry :)), so I tried 230 files.  It worked, but as you can see from the below screenshot the memory was getting quite full, and would probably fail at 300-400 files.  
![image](https://user-images.githubusercontent.com/34637659/36878606-a62a4824-1e24-11e8-8a27-5571946b731e.png)

So after this test, I recieved your message about an update.  I tried the updated package and it dramatically changed the memory burden, did not see the dip, **worked perfectly giving the same results without compromising speed at all.**  See memory usage screen shot below.
![image](https://user-images.githubusercontent.com/34637659/36878697-03f65c04-1e25-11e8-9475-c4df0d8d3a18.png)

So, further good news, I tried the 920 files and it works perfectly, processing the findChromPeaks step in 2hr 40 mins.  Screenshot of memory burden while it was running for completeness of the story below:
![image](https://user-images.githubusercontent.com/34637659/36878723-2adcceb6-1e25-11e8-8372-2ef12d2ee54c.png)

Thanks again for looking into this, i have several other analyses of these 920 samples to process so will continue on and let you know if I stumble across any other problems.

Best Regards
Karl Fraser

**Comment by jotsetung** [comment]
Happy that the fix helped. Thanks for reporting this issue and the detailed descriptions Karl!

To explain what the problem was:
Before performing the chromatographic peak detection I check if the data is centroided and show a warning otherwise. Whether the data is centroided is determined on the first spectrum:
```r
isCentroided(object[[1]])
```
this looks rather harmless, since extracting one spectrum from one file is expected to be really fast and memory efficient using `OnDiskMSnExp` objects. Now, `OnDiskMSnExp` objects are defined in `MSnbase`, along with the `pSet` and `MSnExp` which are this objects parents (`MSnExp` is the *in memory* variant of `OnDiskMSnExp` and is its direct parent). In `xcms` I'm further extending the `OnDiskMSnExp` object with the `XCMSnExp` (inherits all methods from `OnDiskMSnExp`). The problem was that in `xcms` I was, instead of defining a `[[,XCMSnExp` method (that supports handling of adjusted retention times) overwriting the `[[,OnDiskMSnExp` method. The end result was that calling `[[` on an `OnDiskMSnExp`, or `XCMSnExp` object with the `xcms` package being loaded caused the full data being loaded first into memory. In your case that meant that this single `[[1]]` call caused the whole > 900 files being loaded into memory. This memory was never really freed until the end of the function call.

Now this was a lengthy explanation - but happy that we could solve this problem.

**Comment by jotsetung** [comment]
I'm closing the issue as it seems fixed. Feel free to re-open if necessary.

**Comment by tszhonglaw** [comment]
> ```r
> library(devtools)
> install_github("sneumann/xcms", ref = "RELEASE_3_6")
> ```

Hi, Does this only work on macOS? i cannot seem to install - RStudio windows
many thanks,
T

**Comment by stanstrup** [comment]
No. It works everywhere. But you might get dependency issues you have to solve manually (same on all systems). What errors are you getting?

**Comment by tszhonglaw** [comment]
> No. It works everywhere. But you might get dependency issues you have to solve manually (same on all systems). What errors are you getting?

I am getting this msg.... 'Error: Failed to install 'xcms' from GitHub:
  (converted from warning) installation of package ‘C:/Users/thl10/AppData/Local/Temp/RtmpgZxJgH/file1641b523e03/xcms_3.0.2.tar.gz’ had non-zero exit status'

**Comment by stanstrup** [comment]
Is there something above that error? That error seems more a consequence of something upstream.

**Comment by jorainer** [comment]
Ah, yes, I know that error/warning... most likely it is due to a warning that `mzR` was build using a different version of `Rcpp` or similar.

A solution to that seems to use set an environment variable `R_REMOTES_NO_ERRORS_FROM_WARNINGS="true"` (see [here](https://github.com/r-lib/remotes#environment-variables)) - problem is I don't know how one would set an environment variable in Windows :) 

Just one thing - the `RELEASE_3_6` is pretty old! what version of R are you using? With R 3.6 you should install Bioconductor 3.10 (which also contains then this fix), not 3.6 (that's what the `RELEASE_3_6` above means).
Also, if possible, avoid installing from github. Use the recommended way to install packages:

```r
install.packages("BiocManager")
BiocManager::install("xcms")
```
This uses the official Bioconductor installer.

**Comment by stanstrup** [comment]
You can do it through R with:
```r
Sys.setenv(R_REMOTES_NO_ERRORS_FROM_WARNINGS=FALSE)
```
